File: q13-kth-largest/q13-quick-select.js
// LeetCode 215.
// https://leetcode.com/problems/kth-largest-element-in-an-array/

// - Test Cases:
// [5, 6, 4, 2, 3, 1], k = 2   =>  5
// [1, 2, 2, 2, 3, 4], k = 4   =>  2
// [3], k = 1                  =>  3

// - Quick Sort Solution: divide and conquer => (How to find the partition index)
//                 r
// [5, 6, 4, 2, 1, 3]
//  p
//  i   ->   i
//         (2 < 3)

//                     r
// ["2", 6, 4, "5", 1, 3]
//   p
//              i

//                    r
// [2, "3", 4, 5, 1, "6"]
//     "p"
//             i


const swap = function(arr, i, j) {
    let temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
};

const partition = function(arr, left, right) {
    let piv = arr[right];
    let partitionIdx = left;

    for (let i = left; i < right; i++) {
        if (arr[i] < piv) {
            swap(arr, i, partitionIdx);
            partitionIdx++;
        }    
    }
    
    swap(arr, partitionIdx, right);
    return partitionIdx;
};

const quickSelect = function(arr, left, right, idxToFind) {
    if (left < right) {
        let partitionIdx = partition(arr, left, right);
        
        if (idxToFind === partitionIdx)
            return;
        else if (idxToFind < partitionIdx)
            return quickSelect(arr, left, partitionIdx - 1, idxToFind);
        else
            return quickSelect(arr, partitionIdx + 1, right, idxToFind);
    }
};

const findKthLargest = function(nums, k) {
    let idxToFind = nums.length - k;
    quickSelect(nums, 0, nums.length - 1, idxToFind);
    return nums[idxToFind];
};

// O(N + N/2 + N/4 + N/8 + ...) = O(2N)
// time: O(N); 

// worst: O(N^2)
// The worst case is O(N^2) 
// because if the partition element we choose each time 
// ends up being the very last element index, then 
// we actually only reduce our search space by 1 element, 
// which means we get O(N + N - 1 + N - 2 + N - 3 + N - 4....) 
// which is O(N^2)!

// space: O(1) tail recursion

File: ztm-ds-algo/5-recursion/q13-kth-largest/python/q13-quick-select.py
# // LeetCode 215.
# // https://leetcode.com/problems/kth-largest-element-in-an-array/

# tail recursion method
def quickSelect(nums, left, right, target):
    if left < right:
        pivot = partition(nums, left, right)
        if pivot == target:
            return
        elif pivot > target:
            quickSelect(nums, left, pivot - 1, target)
        else:
            quickSelect(nums, pivot + 1, right, target)

def partition(nums, left, right):
    pivot = left
    for i in range(left, right):
        if nums[i] < nums[right]:
            nums[i], nums[pivot] = nums[pivot], nums[i]
            pivot += 1
    nums[right], nums[pivot] = nums[pivot], nums[right]
    return pivot
    
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        target = len(nums) - k
        quickSelect(nums, 0, len(nums) - 1, target)
        return nums[target]


# =========================================================================
# Iterative method

def quickSelect(nums, startIdx, endIdx, position):
    while True:
        pivot, left, right = startIdx, startIdx + 1, endIdx
        while left <= right:
            if nums[left] > nums[pivot] and nums[right] < nums[pivot]:
                nums[left], nums[right] = nums[right], nums[left]
            if nums[left] <= nums[pivot]:
                left += 1
            if nums[right] >= nums[pivot]:
                right -= 1
        nums[pivot], nums[right] = nums[right], nums[pivot]
        
        if right == position:
            return nums[right]
        if right > position:
            endIdx = right - 1
        else:
            startIdx = right + 1
        
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        position = len(nums) - k
        return quickSelect(nums, 0, len(nums) - 1, position)

# Best, Avg -> time: O(n), space: O(1)
# Worst -> time: O(n ^ 2), space: O(1)